libraries/AP_Baro/AP_Baro_ICP201XX.cpp

```diff
@@ -0,0 +1,486 @@
+        goto failed;
+    }
+
+    wait_read();
+
+    dev->set_retries(0);
+
+    instance = _frontend.register_sensor();
+
+    dev->set_device_type(DEVTYPE_BARO_ICP201XX);
+    set_bus_id(instance, dev->get_bus_id());
+
+    dev->get_semaphore()->give();
+
+    dev->register_periodic_callback(CONVERSION_INTERVAL/2, FUNCTOR_BIND_MEMBER(&AP_Baro_ICP201XX::timer, void));
+    return true;
+
+ failed:
+    dev->get_semaphore()->give();
+    return false;
+}
+
+
+void AP_Baro_ICP201XX::dummy_reg()
+{
+    do {
+        uint8_t reg = REG_EMPTY;
+        uint8_t val = 0;
+        dev->transfer(&reg, 1, &val, 1);
+    } while (0);
+}
+
+bool AP_Baro_ICP201XX::read_reg(uint8_t reg, uint8_t *buf, uint8_t len)
+{
+    bool ret;
+    ret = dev->transfer(&reg, 1, buf, len);
+    dummy_reg();
+    return ret;
+}
+
+bool AP_Baro_ICP201XX::read_reg(uint8_t reg, uint8_t *val)
+{
+    return read_reg(reg, val, 1);
+}
+
+bool AP_Baro_ICP201XX::write_reg(uint8_t reg, uint8_t val)
+{
+    bool ret;
+    uint8_t data[2] = { (uint8_t)reg, val };
+    ret = dev->transfer(data, sizeof(data), nullptr, 0);
+    dummy_reg();
+    return ret;
+}
+
+void AP_Baro_ICP201XX::soft_reset()
+{
+    /* Stop the measurement */
+    mode_select(0x00);
+
+    hal.scheduler->delay(2);
+
+    /* Flush FIFO */
+    flush_fifo();
+
+    /* Mask all interrupts */
+    write_reg(REG_FIFO_CONFIG, 0x00);
+    write_reg(REG_INTERRUPT_MASK, 0xFF);
+}
+
+bool AP_Baro_ICP201XX::mode_select(uint8_t mode)
+{
+    uint8_t mode_sync_status = 0;
+
+    do {
+        read_reg(REG_DEVICE_STATUS, &mode_sync_status, 1);
+
+        if (mode_sync_status & 0x01) {
+            break;
+        }
+
+        hal.scheduler->delay(1);
+    } while (1);
+
+    return write_reg(REG_MODE_SELECT, mode);
+}
+
+bool AP_Baro_ICP201XX::read_otp_data(uint8_t addr, uint8_t cmd, uint8_t *val)
+{
+    uint8_t otp_status = 0xFF;
+
+    /* Write the address content and read command */
+    if (!write_reg(REG_OTP_MTP_OTP_ADDR, addr)) {
+        return false;
+    }
+
+    if (!write_reg(REG_OTP_MTP_OTP_CMD, cmd)) {
+        return false;
+    }
+
+    /* Wait for the OTP read to finish Monitor otp_status */
+    do     {
+        read_reg(REG_OTP_MTP_OTP_STATUS, &otp_status);
+
+        if (otp_status == 0) {
+            break;
+        }
+
+        hal.scheduler->delay_microseconds(1);
+    } while (1);
+
+    /* Read the data from register */
+    if (!read_reg(REG_OTP_MTP_RD_DATA, val)) {
+        return false;
+    }
+
+    return true;
+}
+
+bool AP_Baro_ICP201XX::get_sensor_data(float *pressure, float *temperature)
+{
+    uint8_t fifo_data[96] {0};
+    uint8_t fifo_packets = 0;
+    int32_t data_temp = 0;
+    int32_t data_press = 0;
+    *pressure = 0;
+    *temperature = 0;
+
+    if (read_reg(REG_FIFO_FILL, &fifo_packets)) {
+        fifo_packets  = (uint8_t)(fifo_packets & 0x1F);
+        if (fifo_packets > 0 && fifo_packets <= 16 && read_reg(REG_FIFO_BASE, fifo_data, fifo_packets * 2 * 3)) {
```

Comment:
can fifo_packets be over 16? If so it looks like we will stop reading

Reply:
ccording to the datasheet, the fifo_packets maximum value is 16.

Reply:
the datasheet shows this:
![image](https://user-images.githubusercontent.com/831867/175853378-f03035ef-ce95-4718-9ccb-6927257f0df8.png)
the problem is that if it ever does give one of these reserved values then the driver will never continue. I think you either need to call flush_fifo() here if fifo_packets > 16

Reply:
yes